filebeat/inputsource/udp/server.go

```diff
@@ -0,0 +1,92 @@
+package udp
+
+import (
+	"net"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common/atomic"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Config options for the UDPServer
+type Config struct {
+	Host           string        `config:"host"`
+	MaxMessageSize int           `config:"max_message_size" validate:"positive,nonzero"`
+	Timeout        time.Duration `config:"timeout"`
+}
+
+// Server creates a simple UDP Server and listen to a specific host:port and will send any
+// event received to the callback method.
+type Server struct {
+	running  atomic.Bool
+	config   *Config
+	callback func(data []byte, addr net.Addr)
+	listener net.PacketConn
+	log      *logp.Logger
+	wg       sync.WaitGroup
+}
+
+// New returns a new UDPServer instance.
+func New(config *Config, callback func(data []byte, addr net.Addr)) *Server {
+	return &Server{
+		running:  atomic.MakeBool(false),
+		config:   config,
+		callback: callback,
+		log:      logp.NewLogger("udp").With("address", config.Host),
+	}
+}
+
+// Start starts the UDP Server and listen to incoming events.
+func (u *Server) Start() error {
```

Comment:
Typically a start method does not block like this does. It will "start" some kind of worker goroutine. This encapsulates the logic required to manage a goroutine which is good so that every user is not re-implementing the same thing. As a barebones example:

```
func (x *Thing) Start() {
    x.wg.Add(1)
    go func() {
        defer r.wg.Done()
        x.run()
    }() 
}

func (x *Thing) Stop() {
    close(x.done) 
    x.wg.Wait()
}

func (x *Thing) run() {
    for {
        select {
        case <-r.done:
            return
        default:
        }

        // Do a unit work.
    }
}
```

In your case the socket would replace the done channel. And `Start` would initialize listening socket before starting the worker goroutine. This way `Start` can return an error if it can't "listen" and when start returns the Server is guaranteed to be in the "listening" state (which might remove the need to have `IsRunning`).

Reply:
Thanks for the long input, really appreciated and it clarify things.